Userland/DevTools/UserspaceEmulator/SoftFPU.cpp

```diff
@@ -0,0 +1,1744 @@
+    // type, but this is still difficult, so we use our own lookup for that
+    // We set the alignment bytes to all 1's, too, just in case
+    m_storage[index].__high = ~(decltype(m_storage[index].__high))0u;
+    m_reg_is_mmx[index] = true;
+}
+
+ALWAYS_INLINE void SoftFPU::fpu_push(long double value)
+{
+    if (fpu_is_set(7))
+        fpu_set_stack_overflow();
+    m_fpu_stack_top = (m_fpu_stack_top - 1u) % 8;
+
+    fpu_set(0, value);
+}
+
+ALWAYS_INLINE long double SoftFPU::fpu_pop()
+{
+    warn_if_mmx_absolute(m_fpu_stack_top);
+
+    if (!fpu_is_set(0))
+        fpu_set_stack_underflow();
+
+    auto ret = fpu_get(0);
+    fpu_set_tag(0, FPU_Tag::Empty);
+    m_fpu_stack_top = (m_fpu_stack_top + 1u) % 8;
+    return ret;
+}
+
+ALWAYS_INLINE void SoftFPU::fpu_set_exception(FPU_Exception ex)
+{
+    switch (ex) {
+    case FPU_Exception::StackFault:
+        m_fpu_error_stackfault = 1;
+        m_fpu_error_invalid = 1; // Implies InvalidOperation
+        break;
+    case FPU_Exception::InvalidOperation:
+        m_fpu_error_invalid = 1;
+        if (!m_fpu_mask_invalid)
+            break;
+        return;
+    case FPU_Exception::DenormalizedOperand:
+        m_fpu_error_denorm = 1;
+        if (!m_fpu_mask_denorm)
+            break;
+        return;
+    case FPU_Exception::ZeroDivide:
+        m_fpu_error_zero_div = 1;
+        if (!m_fpu_mask_zero_div)
+            break;
+        return;
+    case FPU_Exception::Overflow:
+        m_fpu_error_overflow = 1;
+        if (!m_fpu_mask_overflow)
+            break;
+        return;
+    case FPU_Exception::Underflow:
+        m_fpu_error_underflow = 1;
+        if (!m_fpu_mask_underflow)
+            break;
+        return;
+    case FPU_Exception::Precision:
+        m_fpu_error_precision = 1;
+        if (!m_fpu_mask_precision)
+            break;
+        return;
+    }
+
+    // set exception bit
+    m_fpu_error_summary = 1;
+
+    // FIXME: set traceback
+    // For that we need to get the currently executing instruction and
+    // the previous eip
+
+    // FIXME: Call FPU Exception handler
+    reportln("Trying to call Exception handler from {}", fpu_exception_string(ex));
+    fpu_dump_env();
+    m_emulator.dump_backtrace();
+    TODO();
+}
+
+template<Arithmetic T>
+ALWAYS_INLINE T SoftFPU::fpu_round(long double value) const
+{
+    // FIXME: may need to set indefinite values manually
+    switch (fpu_get_round_mode()) {
+    case RoundingMode::NEAREST:
+        return static_cast<T>(roundl(value));
+    case RoundingMode::DOWN:
+        return static_cast<T>(floorl(value));
+    case RoundingMode::UP:
+        return static_cast<T>(ceill(value));
+    case RoundingMode::TRUNK:
+        return static_cast<T>(truncl(value));
+    default:
+        VERIFY_NOT_REACHED();
+    }
+}
+
+template<Arithmetic T>
+ALWAYS_INLINE T SoftFPU::fpu_round_checked(long double value)
+{
+    T result = fpu_round<T>(value);
+    if (auto rnd = value - result) {
```

Comment:
Prefer in two lines; this smells like a mistake but isn't.

Reply:
fixed the actual behavior, the current one was wrong, c1 should always be cleared